sinks/xray/xray.go

```diff
@@ -0,0 +1,220 @@
+	// we can store it as a uint32 instead of a float64 and compare apples-to-apples
+	// with the output of our hashing algorithm.
+	sampleThreshold = uint32(sampleRatePercentage * math.MaxUint32 / 100)
+
+	// Build a regex for cleaning names based on valid characters from:
+	// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields
+	reg, err := regexp.Compile("[^a-zA-Z0-9_\\.\\:\\/\\%\\&#=+\\-\\@\\s\\\\]+")
+	if err != nil {
+		return nil, err
+	}
+
+	return &XRaySpanSink{
+		daemonAddr:      daemonAddr,
+		sampleThreshold: sampleThreshold,
+		commonTags:      commonTags,
+		log:             log,
+		nameRegex:       reg,
+	}, nil
+}
+
+// Start the sink
+func (x *XRaySpanSink) Start(cl *trace.Client) error {
+	x.traceClient = cl
+
+	xrayDaemon, err := net.ResolveUDPAddr("udp", x.daemonAddr)
+	if err != nil {
+		return err
+	}
+	conn, err := net.DialUDP("udp", nil, xrayDaemon)
+	if err != nil {
+		return err
+	}
+	x.conn = conn
+
+	return nil
+}
+
+// Name returns this sink's name.
+func (x *XRaySpanSink) Name() string {
+	return "xray"
+}
+
+// Ingest takes in a span and passed it along to the X-Ray client after
+// some sanity checks and improvements are made.
+func (x *XRaySpanSink) Ingest(ssfSpan *ssf.SSFSpan) error {
+	if err := protocol.ValidateTrace(ssfSpan); err != nil {
+		return err
+	}
+
+	sampleCheckValue := []byte(strconv.FormatInt(ssfSpan.TraceId, 10))
+	hashKey := crc32.ChecksumIEEE(sampleCheckValue)
+	if hashKey > x.sampleThreshold {
+		atomic.AddInt64(&x.spansDropped, 1)
+		return nil
+	}
+
+	annos := map[string]string{}
+	for k, v := range x.commonTags {
+		annos[k] = v
+	}
+	for k, v := range ssfSpan.Tags {
+		if k != XRayTagNameClientIP {
+			annos[k] = v
+		}
+	}
+
+	name := string(x.nameRegex.ReplaceAll([]byte(ssfSpan.Service), []byte("_")))
+	if len(name) > 200 {
+		name = name[:200]
+	}
+
+	// The fields below are defined here:
+	// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields
+	segment := XRaySegment{
+		// ID is a 64-bit hex
+		ID: fmt.Sprintf("%016x", ssfSpan.Id),
+		// Trace ID is version-startTimeUnixAs8CharHex-traceIdAs24CharHex
+		TraceID:     fmt.Sprintf("1-%08x-%024x", ssfSpan.StartTimestamp/1e9, ssfSpan.TraceId),
+		Name:        name,
+		StartTime:   float64(float64(ssfSpan.StartTimestamp) / float64(time.Second)),
+		EndTime:     float64(float64(ssfSpan.EndTimestamp) / float64(time.Second)),
+		Annotations: annos,
+		Namespace:   "remote",
+		Error:       ssfSpan.Error,
+		// Because X-Ray doesn't offer another way to get this data in, we pretend
+		// it's HTTP for now. It's likely that as X-Ray and/or Veneur develop this
+		// will change.
+		HTTP: XRaySegmentHTTP{
+			Request: XRaySegmentHTTPRequest{
+				URL:      fmt.Sprintf("%s:%s", ssfSpan.Service, ssfSpan.Name),
+				ClientIP: ssfSpan.Tags[XRayTagNameClientIP],
+				// Method:   "GET",
```

Comment:
Are these comments intentional?

Reply:
Sorta? Just to explore the format. Should probably flag why there are here.